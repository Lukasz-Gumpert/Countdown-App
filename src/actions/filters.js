
export const setTextFilter = (text) => ({ 
    type: 'SET_TEXT_FILTER',
    text
});

export const sortByName = () => ({ 
    type: 'SORT_BY_NAME'
});

export const sortByType = () => ({
    type: 'SORT_BY_TYPE'
});

export const endingSoonest = () => ({
    type: 'ENDING_SOONEST'
});

export const furthestAway = () => ({
    type: 'FURTHEST_AWAY'
});


