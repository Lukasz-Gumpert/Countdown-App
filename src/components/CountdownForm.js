import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class CountdownForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            title: '',
            category: 'movies',
            releaseDate: moment(),
            description: '',
            calendarFocused: false,
            error: '',
        };
    };

    onTitleChange = (e) => {
        const title = e.target.value;
        this.setState(() => ({ title }));
        console.log(this.state.title)
    };

    onCategoryChange = (sel) => {
        const category = sel.target.value;
        this.setState(() => ({ category }));
    };

    onDateChange = (releaseDate) => {
        if (releaseDate) {
            this.setState(() => ({ releaseDate }));
        };
    };

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }))
    };

    onSubmit = (e) => {
        e.preventDefault();
        this.props.onSubmit({
            title: this.state.title,
            category: this.state.category,
            releaseDate: this.state.releaseDate.valueOf(),
            description: this.state.description
        });
    };

    render() {
        return (
            <form  onSubmit={this.onSubmit}>
                <input
                    className="text-input"
                    type="text"
                    placeholder="title"
                    autoFocus
                    value={this.state.title}
                    onChange={this.onTitleChange}
                />
                <select
                    value={this.state.category}
                    onChange={this.onCategoryChange}
                    name="category"
                    className="select-category"
                >
                    <option value="movies">Movies</option>
                    <option value="music">Music</option>
                    <option value="tech">Tech</option>
                    <option value="games">Games</option>
                    <option value="events">Events</option>
                </select>
                <SingleDatePicker
                    date={this.state.releaseDate}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                <input
                    className="text-input"
                    type="text"
                    placeholder="description"
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
                <button>Save</button>
            </form>

        )
    }
}