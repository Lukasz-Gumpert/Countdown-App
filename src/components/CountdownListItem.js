import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment'
import CountdownClock from './CountdownClock';

const CountdownListItem = ({ id, title, category, releaseDate, description }) => (
    <Link to={`edit/${id}`} className="list-item">
        <p> {category} </p>
        <h1>{title}</h1> : {description} <p>Time left:</p>
        <CountdownClock timeTillDate={moment(releaseDate)} />
    </Link>
);

export default CountdownListItem;