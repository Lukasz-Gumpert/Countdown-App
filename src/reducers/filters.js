const filtersReducerDefaultState = {
    text: '',
    sortBy: 'sortByName',
};

export default (state = filtersReducerDefaultState, action) => {
    switch (action.type) {
        case 'SET_TEXT_FILTER':
            return {
                ...state, text: action.text
            };
        case 'SORT_BY_NAME':
            return {
                ...state, sortBy: 'sortByName'
            };
        case 'ENDING_SOONEST':
            return {
                ...state, sortBy: 'endingSoonest'
            };
        case 'FURTHEST_AWAY':
            return {
                ...state, sortBy: 'furthestAway'
            };
        case 'SORT_BY_TYPE':
            return {
                ...state, sortBy: 'sortByType'
            };
        default:
            return state
    };
};