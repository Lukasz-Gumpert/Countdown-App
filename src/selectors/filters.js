import moment from 'moment';

const selectFilters = (countdowns, { sortBy, text, releaseDate }) => {
    return countdowns.filter((countdown) => {
        const textMatch = countdown.title.toLowerCase().includes(text.toLowerCase()) || countdown.description.toLowerCase().includes(text.toLowerCase())
        return textMatch
    }).sort((a, b) => {
        if (sortBy === "sortByName") {
            if (a.title == b.title) {
                return (a.releaseDate < b.releaseDate) ? -1 : (a.releaseDate < b.releaseDate) ? 1 : 0;
            } else {
                return (a.title < b.title) ? -1 : 1;
            };
        } else if (sortBy === "endingSoonest") {
            if (a.releaseDate == b.releaseDate) {
                return (a.title < b.title) ? -1 : (a.title < b.title) ? 1 : 0;
            } else {
                return (a.releaseDate < b.releaseDate) ? -1 : 1;
            };
        } else if (sortBy === "furthestAway") {
            if (a.releaseDate == b.releaseDate) {
                return (a.title < b.title) ? -1 : (a.title < b.title) ? 1 : 0;
            } else {
                return (a.releaseDate > b.releaseDate) ? -1 : 1;
            };
        } else if (sortBy === "type") {
            if (a.category == b.category) {
                return (a.releaseDate < b.releaseDate) ? -1 : (a.releaseDate < b.releaseDate) ? 1 : 0;
            } else {
                return (a.category > b.category) ? 1 : -1;
            };
        };
    });
};

// var af = a.title; 
// var bf = b.title; 
// var as = moment(releaseDate); 
// var bs = moment(releaseDate); 

// if(af == bf) { 
//     return (as < bs) ? -1 : (as < bs) ? 1 : 0; 
// } else { 
//     return (af < bf) ? -1 : 1; 
// } 



//         if (sortBy === 'sortByName') {
//             return (a.title === b.title) ? a.title > b.title ? 1 : -1;
//         } else if (sortBy === 'endingSoonest') {
//             return a.releaseDate > b.releaseDate ? 1 : -1;
//         } else if (sortBy === 'furthestAway') {
//             return a.releaseDate < b.releaseDate ? 1 : -1;
//         } else if (sortBy === 'sortByType') {
//             return a.category > b.category ? 1 : -1;
//         };
//     });
// };



// const pipeComparators = (...fns) => (a, b) => {
//     const compare = (a, b, f = 0) =>
//       fns[f](a, b) ? fns[f](a, b) : fns[f](b, a) ? fns[f](b, a) : f !== fns.length - 1 ? compare(a, b, f + 1) : 0;
//     return compare(a, b);
//   };
   
//   const comparator = fn => (a, b) => (fn(a, b) ? 1 : fn(b, a) ? -1 : 0);
   
//   const filters = {
//     sortByName: comparator((a, b) => (a.title > b.title ? 1 : -1)),
//     endingSoonest: comparator((a, b) => (a.releaseDate > b.releaseDate ? 1 : -1)),
//     furthestAway: comparator((a, b) => (a.releaseDate < b.releaseDate ? 1 : -1)),
//     sortByType: comparator((a, b) => (a.category > b.category ? 1 : -1)),
//   };
   
//   const sortByProperty = propName => comparator((a, b) => (a[propName] > b[propName] ? 1 : -1));
   
//   const getFilterByName = filterName => {
//     const filter = filters[filterName];
//     return filter || comparator(() => 0);
//   };
   
//   const selectFilters = (countdowns, { sortBy, text, releaseDate }) => {
//     return countdowns
//       .filter(
//         ({ title, description }) =>
//           title.toLowerCase().includes(text.toLowerCase()) || description.toLowerCase().includes(text.toLowerCase()),
//       )
//       .sort(pipeComparators(sortByProperty('TU_NAZWA_WLASCIWOSCI'), getFilterByName(sortBy)));

export default selectFilters;