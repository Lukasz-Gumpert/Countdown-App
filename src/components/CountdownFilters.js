import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByName, sortByType, endingSoonest, furthestAway } from '../actions/filters';

export class CountdownFilters extends React.Component {
    constructor(props) {
        super(props);

        this.onTextChange = (e) => {
            return this.props.setTextFilter(e.target.value);
        }

        this.onSortChange = (sel) => {
            const sortBy = sel.target.value;
            if (sortBy === "sortByName") {
                return props.sortByName();
            } else if (sortBy === "sortByType") {
                return props.sortByType();
            } else if (sortBy === "endingSoonest") {
                return props.endingSoonest();
            } else if (sortBy === "furthestAway") {
                return props.furthestAway();
            };
        };
    };

    render() {
        return (
            <div>
                <input
                    type="text"
                    placeholder="search"
                    value={this.props.filters.text}
                    onChange={this.onTextChange}
                />
                <select
                    value={this.props.filters.sortBy}
                    onChange={this.onSortChange}
                    name="sortBy"
                >
                    <option value="endingSoonest">Ending soonest</option>
                    <option value="furthestAway" >Furthest away</option>
                    <option value="sortByName" >By name</option>
                    <option value="sortByType" >By type</option>
                </select>
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (data) => dispatch(setTextFilter(data)),
    sortByName: () => dispatch(sortByName()),
    sortByType: () => dispatch(sortByType()),
    endingSoonest: () => dispatch(endingSoonest()),
    furthestAway: () => dispatch(furthestAway()),
});


export default connect(mapStateToProps, mapDispatchToProps)(CountdownFilters);