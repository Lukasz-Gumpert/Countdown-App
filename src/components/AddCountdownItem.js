import React from 'react';
import { connect } from 'react-redux';
import CountdownForm from './CountdownForm';
import { startAddCountdown } from '../actions/countdowns';
import moment from 'moment';

export class AddCountdownItem extends React.Component {
    onSubmit = (countdown) => {
        this.props.startAddCountdown(countdown);
        this.props.history.push('/');

    }
 
    render() {
        return (
            <div>
                <h1>Add countdown</h1>
                <CountdownForm onSubmit={this.onSubmit} />
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => ({
    startAddCountdown: (countdown) => dispatch(startAddCountdown(countdown))
});

export default connect(undefined, mapDispatchToProps)(AddCountdownItem);
