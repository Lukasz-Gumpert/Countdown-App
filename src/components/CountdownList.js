import React from 'react';
import { connect } from "react-redux";
import CountdownListItem from '../components/CountdownListItem';
import selectFilters from '../selectors/filters';


export const CountdownList = (props) => (
    <div>
        {
            props.countdowns.length === 0 ? (
                <p>No countdown data</p>
            ) : (
                    props.countdowns.map((countdown) => {
                        return <CountdownListItem key={countdown.id} {...countdown} />
                    })
                )
        }
    </div>
);

const mapStateToProps = (state) => {
    return {
        countdowns: selectFilters(state.countdowns, state.filters)
    }
};

export default connect(mapStateToProps)(CountdownList);