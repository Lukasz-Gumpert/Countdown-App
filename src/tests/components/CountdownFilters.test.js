import React from 'react';
import { shallow } from 'enzyme';
import { CountdownFilters } from '../../components/CountdownFilters';
import { filters, altFilters } from '../fixtures/filters';

let wrapper, setTextFilter, sortByName, sortByType, endingSoonest, furthestAway

beforeEach(() => {
    setTextFilter = jest.fn();
    sortByName = jest.fn();
    sortByType = jest.fn();
    endingSoonest = jest.fn();
    furthestAway = jest.fn();
    wrapper = shallow(
        <CountdownFilters
            filters={filters}
            setTextFilter={setTextFilter}
            sortByName={sortByName}
            sortByType={sortByType}
            endingSoonest={endingSoonest}
            furthestAway={furthestAway}
        />);
});

test('Should render CountdownFilters correctly', () => {
    expect(wrapper).toMatchSnapshot()
});

test('should handle text change', () => {
    const value = 'rent';
    wrapper.find('input').simulate('change', { target: { value } });
    expect(setTextFilter).toHaveBeenLastCalledWith(value);
});

test('Should handle sort by name', () => {
    const value = "sortByName"
    wrapper.setProps({ filters: altFilters });
    wrapper.find('select').simulate('change', { target: {value}}  );
    expect(sortByName).toHaveBeenCalled();
});

test('Should handle sort by type', () => {
    const value = "sortByType"
    wrapper.setProps({ filters: altFilters });
    wrapper.find('select').simulate('change', { target: {value}}  );
    expect(sortByType).toHaveBeenCalled();
});

test('Should handle sort by ending soonest', () => {
    const value = "endingSoonest"
    wrapper.setProps({ filters: altFilters });
    wrapper.find('select').simulate('change', { target: {value}}  );
    expect(endingSoonest).toHaveBeenCalled();
});

test('Should handle sort by furthest away', () => {
    const value = "furthestAway"
    wrapper.setProps({ filters: altFilters });
    wrapper.find('select').simulate('change', { target: {value}}  );
    expect(furthestAway).toHaveBeenCalled();
});