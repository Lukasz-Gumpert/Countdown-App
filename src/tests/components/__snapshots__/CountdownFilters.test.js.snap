// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render CountdownFilters correctly 1`] = `
<div>
  <input
    onChange={[Function]}
    placeholder="search"
    type="text"
    value=""
  />
  <select
    name="sortBy"
    onChange={[Function]}
    value="title"
  >
    <option
      value="endingSoonest"
    >
      Ending soonest
    </option>
    <option
      value="furthestAway"
    >
      Furthest away
    </option>
    <option
      value="sortByName"
    >
      By name
    </option>
    <option
      value="sortByType"
    >
      By type
    </option>
  </select>
</div>
`;
